### PR TITLE
Add Quanta OEM code: ipmitool for DIMM slot readable feature

### PR DESCRIPTION
**Pull request by qctbmc**
Hi, Mr. Amelkin:

I posted ticket with https://sourceforge.net/p/ipmitool/bugs/511/ on SourceForge.
I have forked ipmitool and try "Pull Request" to request merge to "ipmitool" code.

Thank you

### CODE REVIEW

File: lib/ipmi_sel.c [hunk 1]
```diff
@@ -703,6 +703,82 @@ get_supermicro_evt_desc(struct ipmi_intf *intf, struct sel_event_record *rec)
...
+			if (rsp == NULL) {
+				lprintf(LOG_ERR, " Error getting system info");
+				if (desc != NULL) {
+					free(desc);
+					desc = NULL;
+				}
+				return NULL;
+			} else if (rsp->ccode > 0) {
+				lprintf(LOG_ERR, " Error getting system info: %s",
+						val2str(rsp->ccode, completion_code_vals));
```
Comment (AlexanderAmelkin) [inline lib/ipmi_sel.c]: Please use tabs to indent to `lprintf` level, then spaces to align `val2str` to `LOG_ERR` or to the opening double quote. Please see https://github.com/ipmitool/ipmitool/wiki/Coding-Standards regarding indentation and alignment.

File: lib/ipmi_sel.c [hunk 2]
```diff
@@ -703,6 +703,82 @@ get_supermicro_evt_desc(struct ipmi_intf *intf, struct sel_event_record *rec)
...
+			}
+			/* check the platform type */
+			oem_id = ipmi_get_qct_paltform_id(intf);
+						
+			if(oem_id == 0x02)
+			{
+				snprintf(desc, SIZE_OF_DESC, "CPU%d_%c%d",
+						 (data3 & 0xC0)>> 6,
+						 ((data3 & 0x38) >> 3)+0x41,						 
+						 (data3 & 0x07));
```
Comment (AlexanderAmelkin) [inline lib/ipmi_sel.c]: Please indent/align according to https://github.com/ipmitool/ipmitool/wiki/Coding-Standards

File: lib/helper.c [hunk 3]
```diff
@@ -1026,3 +1026,40 @@ ipmi_get_oem_id(struct ipmi_intf *intf)
 
 	return oem_id;
 }
+
+uint16_t
+ipmi_get_qct_paltform_id(struct ipmi_intf *intf)
```
Comment (AlexanderAmelkin) [inline lib/helper.c]: This doesn't belong here. It must go to `lib/ipmi_quantaoem.c` just like it is done for sunoem, kontronoem, et al.

File: include/ipmitool/ipmi.h [hunk 4]
```diff
@@ -250,6 +250,7 @@ struct ipmi_rs {
 #define IPMI_BMC_SLAVE_ADDR		0x20
 #define IPMI_REMOTE_SWID		0x81
 
+#define IPMI_NETFN_QCTOEM		0x36
```
Comment (AlexanderAmelkin) [inline include/ipmitool/ipmi.h]: This doesn't belong here and must go to `include/ipmitool/ipmi_quantaoem.h` just like it is done for sunoem, et al.

**Comment by qctbmc** [comment]
Hi, Mr. Amelkin:

Thanks for your advice, I have fixed the typo and move Quanta's code to new files.
Please let me know if my code need to be modified.

Thanks

File: lib/ipmi_sel.c [hunk 5]
```diff
@@ -1244,6 +1244,9 @@ ipmi_get_oem_desc(struct ipmi_intf * intf, struct sel_event_record * rec)
 	case IPMI_OEM_SUPERMICRO_47488:
 		desc = get_supermicro_evt_desc(intf, rec);
 		break;
+	case IPMI_OEM_QUANTA:		
+		desc = get_quanta_evt_desc(intf, rec);
```
Comment (AlexanderAmelkin) [inline lib/ipmi_sel.c]: You get `warning: implicit declaration of function ‘get_quanta_evt_desc’` here, and also (as a result) `warning: implicit declaration of function ‘get_quanta_evt_desc’`. Please fix.

File: lib/ipmi_quantaoem.c [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+{
+	struct ipmi_rs *rsp;
+	struct ipmi_rq req;
+	char *desc = NULL;
+	int chipset_type = 0;
+	int data1;
+	int data2;
+	int data3;
+	int sensor_type;
+	uint8_t i = 0;
```
Comment (AlexanderAmelkin) [inline lib/ipmi_quantaoem.c]: This variable is unused.

File: lib/ipmi_quantaoem.c [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+			}
+			/* check the platform type */
+			oem_id = ipmi_get_qct_platform_id(intf);
+						
+			if(oem_id == 0x02)
+			{
+				snprintf(desc, SIZE_OF_DESC, "CPU%d_%c%d",
+				         (data3 & 0xC0)>> 6,
+				         ((data3 & 0x38) >> 3)+0x41,						 
+				         (data3 & 0x07));
```
Comment (AlexanderAmelkin) [inline lib/ipmi_quantaoem.c]: What are all these magic numbers? Macros with descriptive names are required.

File: lib/ipmi_quantaoem.c [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+				if (desc != NULL) {
+					free(desc);
+					desc = NULL;
+				}
+				return NULL;
+			}
+			/* check the platform type */
+			oem_id = ipmi_get_qct_platform_id(intf);
+						
+			if(oem_id == 0x02)
```
Comment (AlexanderAmelkin) [inline lib/ipmi_quantaoem.c]: What is `0x02` ? A descriptive macro please.

File: lib/ipmi_quantaoem.c [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+	return oem_id;
+}
+
+char *
+get_quanta_evt_desc(struct ipmi_intf *intf, struct sel_event_record *rec)
+{
+	struct ipmi_rs *rsp;
+	struct ipmi_rq req;
+	char *desc = NULL;
+	int chipset_type = 0;
```
Comment (AlexanderAmelkin) [inline lib/ipmi_quantaoem.c]: This variable is never used.

**Comment by AlexanderAmelkin** [comment]
@qctbmc, will you please fix it? It's out of sync with master already and the longer you wait, the more out of sync it will get.

**Comment by qctbmc** [comment]
Hi, Mr. Amelkin:

Sorry to reply late, I already sync with master code and modify Quanta's code from your comment.
Thanks for your advice, please let me know if my code need to be modified.

Thanks

File: lib/ipmi_quantaoem.c [hunk 10]
```diff
@@ -93,10 +96,10 @@ ipmi_get_qct_platform_id(struct ipmi_intf *intf)
 		lprintf(LOG_ERR, "Get Platform ID command failed: %#x %s", rsp->ccode, val2str(rsp->ccode, completion_code_vals));
 		return 0;
 	}
-	oem_id = rsp->data[0];
-	lprintf(LOG_DEBUG,"Platform ID: %x", oem_id);
+	platform_id = rsp->data[0];
+	lprintf(LOG_DEBUG,"Platform ID: %x", platform_id);
 
-	return oem_id;
+	return platform_id;
```
Comment (AlexanderAmelkin) [inline lib/ipmi_quantaoem.c]: You're still returning an unsigned byte up-converted to unsigned int16. What's the purpose?
Besides, in fact you're apparently returning your freshly defined enum type, `oem_qct_platform_id_t` (or `Code_Base` in your committed version). Please change the function type to that. Also, the `0` that you return above to indicate an error seems to be a valid platform_id. Namely, it's GRANTLEY. So you can't use 0. You need to define some `OEM_QCT_PLATFORM_UNKNOWN` value in your enum and return it.

Please pay more attention when processing the review comments.

File: lib/ipmi_quantaoem.c [hunk 11]
```diff
@@ -1,33 +1,34 @@
 /*
- * Copyright (c) 2009, 2014, Oracle and/or its affiliates. All rights reserved.
- *
+ * Copyright (c) 2018 Quanta Computer Inc. All Rights Reserved.
+ *
```
Comment (AlexanderAmelkin) [inline lib/ipmi_quantaoem.c]: This trailing space (and the one below) must not be here.

File: include/ipmitool/ipmi_quantaoem.h [hunk 12]
```diff
@@ -12,22 +12,23 @@
...
- * OR FOR DIRECT, INDIRECT, SPECIAL, CONSEQUENTIAL, INCIDENTAL OR
- * PUNITIVE DAMAGES, HOWEVER CAUSED AND REGARDLESS OF THE THEORY OF
- * LIABILITY, ARISING OUT OF THE USE OF OR INABILITY TO USE THIS SOFTWARE,
- * EVEN IF SUN HAS BEEN ADVISED OF THE POSSIBILITY OF SUCH DAMAGES.
+ * Quanta Computer Inc. OR ITS LICENSORS BE LIABLE FOR ANY LOST REVENUE, 
+ * PROFIT OR DATA, OR FOR DIRECT, INDIRECT, SPECIAL, CONSEQUENTIAL, 
+ * INCIDENTAL OR PUNITIVE DAMAGES, HOWEVER CAUSED AND REGARDLESS OF THE 
+ * THEORY OF LIABILITY, ARISING OUT OF THE USE OF OR INABILITY TO USE 
+ * THIS SOFTWARE, EVEN IF SUN HAS BEEN ADVISED OF THE POSSIBILITY OF 
+ * SUCH DAMAGES.
```
Comment (AlexanderAmelkin) [inline include/ipmitool/ipmi_quantaoem.h]: Please in future avoid such re-formatting as they appear like major changes when in fact you only moved the line breaks.

Please remove the trailing whitespaces in this new block.

Please make sure your committed code doesn't contain traling whitespaces (including tabs) anywhere.

File: lib/ipmi_quantaoem.c [hunk 13]
```diff
@@ -1,33 +1,34 @@
 /*
- * Copyright (c) 2009, 2014, Oracle and/or its affiliates. All rights reserved.
- *
+ * Copyright (c) 2018 Quanta Computer Inc. All Rights Reserved.
+ * 
  * Redistribution and use in source and binary forms, with or without
  * modification, are permitted provided that the following conditions
  * are met:
- *
+ *
```
Comment (AlexanderAmelkin) [inline lib/ipmi_quantaoem.c]: Trailing whitespace again. Here and below.

File: lib/ipmi_quantaoem.c [hunk 14]
```diff
@@ -80,7 +83,7 @@ ipmi_get_qct_platform_id(struct ipmi_intf *intf)
 	
 	memset(&req, 0, sizeof(req));
 	req.msg.netfn = IPMI_NETFN_QCTOEM;
-	req.msg.cmd   = 0x65;	
+	req.msg.cmd   = BMC_GET_OEM_INFO;
```
Comment (AlexanderAmelkin) [inline lib/ipmi_quantaoem.c]: Trailing whitespace here.

File: lib/ipmi_quantaoem.c [hunk 15]
```diff
@@ -157,20 +152,21 @@ get_quanta_evt_desc(struct ipmi_intf *intf, struct sel_event_record *rec)
...
+			platform_id = ipmi_get_qct_platform_id(intf);
 						
-			if(oem_id == 0x02)
+			if(platform_id == PURLEY)
 			{
 				snprintf(desc, SIZE_OF_DESC, "CPU%d_%c%d",
-				         (data3 & 0xC0)>> 6,
-				         ((data3 & 0x38) >> 3)+0x41,						 
-				         (data3 & 0x07));
+						 CPU_NUM( data ),
```
Comment (AlexanderAmelkin) [inline lib/ipmi_quantaoem.c]: Trailing whitespace

**Comment by AlexanderAmelkin** [comment]
Also, your 'Update from ipmitool master' commit is not the right way to rebase to master.
To properly rebase use this on your branch, do the following:

```
git fetch ipmitool/master # provided that you have a remote called 'ipmitool'
git co master # this is your branch you requested to pull from
git rebase -i ipmitool/master
```
During the interactive rebase, drop your 'Update' commit, then resolve all
the conflicts and finish rebasing, then forcecefully update your branch:

```
git push -f
```

Test that you can do `git merge --ff-only` on `ipmitool/master` from your `master` branch:

```
git co -b usptream-master ipmitool/master
git merge --ff-only master
```

**Comment by qctbmc** [comment]
Hi, Mr. Amelkin:

Thanks for your great advice, I re-sync master ipmitool code and add my modify code.
Please let me know if my code need to be modified.

Thanks

**Comment by qctbmc** [comment]
Hi, Mr. Amelkin:

Thanks for your great advice, I re-sync master ipmitool code and add my modify code.
Please let me know if my code need to be modified.

Thanks

**Comment by AlexanderAmelkin** [comment]
Ok, I have fixed some more minor issues and merged this.

**Comment by JPlanche** [comment]
I am very sceptical about this patch, it gives me wrong DIMM locations for UE/CE (vs. the technical PDF QCT support gave me).
In fact I think it should not be used as is.

**Comment by AlexanderAmelkin** [comment]
@JPlanche, well, I believe the code was sumbitted by Quanta Cloud Technology BMC team, so I assumed they know what they do about their products. If you think there is a bug, please submit your own PR with a fix or at least create a separate issue. Commenting a merged PR is quite useless in terms of getting the problem fixed. You may reference this PR from your issue or new PR if you want though.

**Pull request merged**